data_structures/binary_search_tree.ts

```diff
@@ -0,0 +1,199 @@
+/**
+ * Represents a node of a binary search tree.
+ *
+ * @template T The type of the value stored in the node.
+ */
+class TreeNode<T> {
+  constructor(
+    public data: T,
+    public leftChild?: TreeNode<T>,
+    public rightChild?: TreeNode<T>,
+  ) {}
+}
+
+/**
+ * An implementation of a binary search tree.
+ *
+ * A binary tree is a tree with only two children per node. A binary search tree on top sorts the children according
+ * to following rules:
+ * - left child < parent node
+ * - right child > parent node
+ * - all children on the left side < root node
+ * - all children on the right side > root node
+ *
+ * For profound information about trees
+ * @see https://www.geeksforgeeks.org/introduction-to-tree-data-structure-and-algorithm-tutorials/
+ *
+ * @template T The data type of the values in the binary tree.
+ */
+export class BinarySearchTree<T> {
+  rootNode?: TreeNode<T>;
+
+  /**
+   * Instantiates the binary search tree.
+   *
+   * @param rootNode The root node.
+   */
+  constructor() {
+    this.rootNode = undefined;
+  }
+
+  /**
+   * Checks, if the binary search tree is empty, i. e. has no root node.
+   *
+   * @returns Whether the binary search tree is empty.
+   */
+  isEmpty(): boolean {
+    return this.rootNode === undefined;
+  }
+
+  /**
+   * Searches for data in the current binary search tree.
+   *
+   * @param data The data to search for.
+   */
+  search(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      throw new Error('No root node defined.');
+    }
+
+    let currentNode = this.rootNode;
+    while (currentNode.data !== data) {
+      if (data > currentNode.data) {
+        if (!currentNode.rightChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.rightChild;
+      } else {
+        if (!currentNode.leftChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.leftChild;
+      }
+    }
+
+    return currentNode;
+  }
+
+  /**
+   * Inserts the given data into the binary search tree.
+   *
+   * @param data The data to be stored in the binary search tree.
+   * @returns
+   */
+  insert(data: T): TreeNode<T> {
```

Comment:
I don't think `insert` should return the new root. It should mutate the tree and return nothing.